### PR TITLE
Read foreign key constraints from the database for use in default load

### PR DESCRIPTION
**Pull request by r-ash**
This PR will
* Add functions to get foreign key constraints from postgres and sqlite databases
* Remove configuration for setting up foreign key constraints
* Use automatic reading of foreign key constraints in default load instead of configured constraints

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## 0.0.3
+
+Read foreign key constraints automatically from the database
```
Comment (richfitz) [inline NEWS.md]: It would be nice to add a (VIMC-2859) here I think so we can relate features back to the original ticket

File: R/foreign_keys.R [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+#'
+#' @return A list representing foreign key constraints
+#' @keywords internal
+get_fk_constraints <- function(con) {
+  dialect <- sql_dialect(con)
+  constraints <- switch(
+    dialect,
+    "sqlite" = get_sqlite_fk(con),
+    "postgresql" = get_postgres_fk(con, dialect),
+    {
```
Comment (richfitz) [inline R/foreign_keys.R]: you don't need curlys here for this (and it's more typical not to use them)

File: R/foreign_keys.R [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+#'
+#' @param con Connection to get constraints for.
+#'
+#' @return A list representing foreign key constraints
+#' @keywords internal
+get_fk_constraints <- function(con) {
+  dialect <- sql_dialect(con)
+  constraints <- switch(
+    dialect,
+    "sqlite" = get_sqlite_fk(con),
```
Comment (richfitz) [inline R/foreign_keys.R]: If you drop the dialect argument from `get_postgres_fk` (as it's a constant in this case) you can write:

```
get_fk <- switch(dialect,
                 postgresql = get_postgres_fk
                 sqlite = get_postgres_fk,
                 stop(sprintf(...)))
constraints <- get_fk(con)
```

File: R/foreign_keys.R [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+#' Get FK constraints for SQLite connection
+#'
+#' @param con SQLite connection to get constraints for.
+#'
+#' @return A data frame representing foreign key constraints.
+#' @keywords internal
+get_sqlite_fk <- function(con) {
+  tables <- DBI::dbListTables(con)
+
+  constraints <- data.frame(matrix(ncol = 4, nrow = 0),
```
Comment (richfitz) [inline R/foreign_keys.R]:
```
data_frame(constraint_table = character(0), constraint_column = character(0), ...)
```

etc?

File: R/foreign_keys.R [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+#' Get FK constraints for postgres connection
+#'
+#' @param con Postgres connection to get constraints for.
+#'
+#' @return A data frame representing foreign key constraints.
+#' @keywords internal
+get_postgres_fk <- function(con) {
+  path <- system.file("sql", "postgresql", "get_fk.sql", package = "dettl")
```
Comment (richfitz) [inline R/foreign_keys.R]: Just a suggestion, but if you're getting a lot of files from the package, you might want to write a helper function:

```r
dettl_file <- function(...) {
  system.file(..., package = "dettl", mustWork = TRUE)
}
```

which you'd use above as:

```r
dettl_file("sql", "postgresql", "get_fk.sql")
```

(the `mustWork` argument errors if not found, which catches a number of bugs)

File: inst/sql/postgresql/get_fk.sql [hunk 6]
```diff
@@ -0,0 +1,24 @@
+WITH unnested_confkey AS (
```
Comment (richfitz) [inline inst/sql/postgresql/get_fk.sql:1]: nice job working all this out :hand:

File: R/foreign_keys.R [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        pragma.'to' as referenced_column,
+        '%s' as constraint_table
+        FROM  pragma_foreign_key_list('%s') as pragma",
+      table, table)
+  })
+
+  constraints <- lapply(queries, function(query) {
+    DBI::dbGetQuery(con, query)
+  })
+  as.data.frame(do.call(rbind, constraints))
```
Comment (richfitz) [inline R/foreign_keys.R]: is the `as.data.frame` needed or wanted here? Each of the result of `dbGetQuery` will be a df so rbinding these together will give you a df

File: R/helper_testing.R [hunk 8]
```diff
@@ -60,10 +60,66 @@ prepare_example_db <- function(dir, add_data = FALSE, add_job_table = FALSE,
...
+#'
+#' @keywords internal
+add_fk_data <- function(con) {
+  DBI::dbExecute(con,
+    "CREATE TABLE region (
+      name TEXT PRIMARY KEY,
+      parent TEXT,
+      FOREIGN KEY (parent) REFERENCES region(name)
+    )")
+    region1 <- data.frame(list(
```
Comment (richfitz) [inline R/helper_testing.R]: what is the `list()` call doing here? (& other `data.frame(list(...))` constructions throughout)

File: tests/testthat/example_default_load/dettl.yml [hunk 9]
```diff
@@ -15,9 +15,3 @@ load:
   default: TRUE
```
Comment (richfitz) [inline tests/testthat/example_default_load/dettl.yml:15]: Not necessarily for this PR, but I wonder if "automatic" is better than "default" as a name for this

**Review by richfitz — COMMENTED** [review]
Super minor things in here - this looks really good and I'm super keen to see how it goes in practice.

I think that this would warrant it's own separate blog post and/or vignette once we get an example - this has always felt like a real pain point with SQL and I think people will be really interested in it - **but** we need a more obvious way of getting it across, as was apparent at standup

**Review by richfitz — CHANGES_REQUESTED** [review]
Nice job on this - super minor comments only here.

I think that this would warrant it's own separate blog post and/or vignette once we get an example - this has always felt like a real pain point with SQL and I think people will be really interested in it - **but** we need a more obvious way of getting it across, as was apparent at standup

File: R/foreign_keys.R (hunk 3 above)
Comment (richfitz) [reply R/foreign_keys.R]: (this was left over from our discussion before)

File: R/foreign_keys.R (hunk 4 above)
Comment (r-ash) [reply R/foreign_keys.R]: Much nicer!

File: R/foreign_keys.R (hunk 2 above)
Comment (r-ash) [reply R/foreign_keys.R]: Got rid of those.

File: NEWS.md (hunk 1 above)
Comment (r-ash) [reply NEWS.md]: I've added that

File: R/foreign_keys.R (hunk 5 above)
Comment (r-ash) [reply R/foreign_keys.R]: That will definitely tidy stuff up a little, I've added that.

File: R/foreign_keys.R (hunk 7 above)
Comment (r-ash) [reply R/foreign_keys.R]: Not needed, I didn't try to see what would happen without the cast. I've removed it now.

File: R/helper_testing.R (hunk 8 above)
Comment (r-ash) [reply R/helper_testing.R]: Not necessary again, I've removed those too.

File: tests/testthat/example_default_load/dettl.yml (hunk 9 above)
Comment (r-ash) [reply tests/testthat/example_default_load/dettl.yml:15]: Yeah agreed I think that is a better name for what it is doing

File: R/dettl.R [hunk 10]
```diff
@@ -15,8 +15,8 @@
 ##'
 ##' @examples
 ##' path <- dettl:::prepare_test_import(
-##'   system.file("examples", "person_information", package = "dettl"),
-##'   system.file("examples", "dettl_config.yml", package = "dettl"))
+##'   dettl:::dettl_file("examples", "person_information"),
```
Comment (richfitz) [inline R/dettl.R]: The examples should continue to use `system.file` - they should use only public API features

**Review by richfitz — CHANGES_REQUESTED** [review]
One tiny thing sorry, then we'll merge this

File: R/dettl.R (hunk 10 above)
Comment (r-ash) [reply R/dettl.R]: I see, I should be exporting `prepare_test_import` then too?

Comment (r-ash) [reply R/dettl.R]: I've updated that back to using `system.file`. I've also gone through the examples and made sure I'm only using exported functions.

**Review by richfitz — APPROVED** [review]

**Pull request merged**